enhancements/installer/aws-custom-dns.md

```diff
@@ -0,0 +1,544 @@
+cloud due to regulatory [ITAR](https://www.gov-relations.com/) or operational
+contraints. It is important for OpenShift to provide a way for these customers
+to use their own preferred DNS solution while supporting their IPI deployments.
+
+This external Custom DNS solution would be responsible for DNS records for
+`api.<cluster domain>` and `*.apps.<cluster domain>`. OpenShift would be
+responsible for in-cluster resolution of `api.<cluster domain>`,
+`api-int.<cluster domain>` and `*.apps.<cluster domain>`.
+The cloud Load Balancer(LB) is still the preferred solution for providing LB
+services for `api`, `api-int` and `ingress` in public and private hosted zones.
+
+### User Stories
+
+- As an administrator, I would like to deploy OpenShift 4 clusters to supported
+cloud providers leveraging my custom DNS service. While cloud-based DNS services
+provide convenient hostname management, there are a number of regulatory (ITAR)
+and operational constraints that prohibit the use of those DNS hosting services
+on public cloud providers.
+
+### Goals
+
+- Enable AWS customers to use their custom DNS solution instead of the cloud
+solution (Route53).
+- Lay the groundwork to add this capability for other cloud platforms like
+Azure and GCP.
+- Continue using the cloud based LB service for API, Ingress and API-Int. 
+
+### Non-Goals
+
+- We will not focus on Installer changes to support the custom DNS solution on
+Azure and GCP platforms. That will come in follow-up work and will rely on the
+changes introduced within this solution.
+
+- We will not use the [Kubernetes External DNS provider](https://github.com/openshift/external-dns)
+and/or the [OpenShift External DNS Operator](https://github.com/openshift/external-dns-operator)
+to program dns entries across varied DNS providers. The External DNS Operator
+relies on being able to program the customer's DNS solution. The customer DNS
+solutions in consideration here do not allow OpenShift to configure them.
+
+## Proposal
+
+For the OpenShift installation to make use of a custom DNS solution that is not
+configurable by OpenShift, the customer is responsible for configuring the
+DNS entries for the API and Ingress services.
+
+When it is indicated to the OpenShift installer via an input parameter that the
+custom DNS solution is in use, OpenShift does not configure the cloud DNS
+solution [Route53 for AWS](https://aws.amazon.com/route53/). OpenShift will
+continue to configure and manage the cloud Load Balancing(LB) solution [Elastic
+Load Balancer](https://aws.amazon.com/elasticloadbalancing/) for all its
+services.
+
+The Installer configures the LBs for the API and API-Int services and the
+Ingress Controllers configure the LBs for the *.apps service. There is
+currently no way of knowing these LB IP addresses ahead of time and so the
+the customer's DNS solution cannot be configured before cluster installation.
+
+OpenShift will start static CoreDNS pods to provide DNS resolution for API,
```

Comment:
In order for a node to come online, they make a call to the machine config server via the api-int domain.  This happens for workers and masters, but not for bootstrap.  

How will a node go through ignition if there's no resolution of the api-int domain it tries to reach out to.  The machine config server will have to be overridden with the DNS label set on the load balancer in AWS. 

In Azure, you can only set dns labels on public IP addresses for this.  Since api-int sits behind a load balancer with a dynamic frontend private IP address, it would have to be passed to the machine config server once the bootstrap node came online.

Reply:
@bennerv Thanks for your comment. If the machine config operator were to start a CoreDNS pod to provide api and api-int resolution (to the LB DNS name) starting from the bootstrap node and then later on the control plane nodes, would that take care of the api-int resolution for node ?

I am working on another proposal where the customer would have to pre-create the api and api-int LBs and configure the custom DNS to resolve these URLs before starting cluster installation.

Reply:
@sadasu Here's the issue.

The `igntionHost` parameter defined always points to [api-int](https://github.com/openshift/installer/blob/master/pkg/asset/ignition/machine/node.go#L29-L30).  Since we're requiring the resolvability of api-int to exist in order for any master or worker to come online, we cannot rely on the resolvability of api-int as a domain itself during cluster creation.  Other platforms get around it by allowing a VIP to be created in advanced, similar to your suggestion where things are precreated.  

If resources are not pre-created, then this method will not work on Azure without feeding the IP address of the internal load balancer back to the `ignitionHost` variable.